lib/common/service-object.js

```diff
@@ -0,0 +1,261 @@
+ * object," like a BigQuery dataset or Storage bucket.
+ *
+ * Most of the time, these objects share common functionality; they can be
+ * created or deleted, and you can get or set their metadata.
+ *
+ * By inheriting from this class, a service object will be extended with these
+ * shared behaviors. Note that any method can be overridden when the service
+ * object requires specific behavior.
+ *
+ * @private
+ *
+ * @param {object} config - Configuration object.
+ * @param {string} config.baseUrl - The base URL to make API requests to.
+ * @param {string} config.createMethod - The method which creates this object.
+ * @param {string} config.id - The identifier of the object. For example, the
+ *     name of a Storage bucket or Pub/Sub topic.
+ * @param {string[]=} config.inherit - If all of the methods don't apply to the
+ *     service object, this is a whitelist of methods to implement.
+ * @param {object} config.parent - The parent service instance. For example, an
+ *     instance of Storage if the object is Bucket.
+ */
+function ServiceObject(config) {
+  var self = this;
+
+  this.metadata = {};
+
+  this.baseUrl = config.baseUrl;
+  this.parent = config.parent; // Parent class.
+  this.id = config.id; // Name or ID (e.g. dataset ID, bucket name, etc.)
+  this.createMethod = config.createMethod;
+
+  if (config.exclude) {
+    var allMethods = Object.keys(ServiceObject.prototype);
+    allMethods.forEach(function(methodName) {
+      if (config.exclude.indexOf(methodName) > -1) {
+        self[methodName] = undefined;
+      }
+    });
+  }
+}
+
+/**
+ * Create the object.
+ *
+ * @param {object=} options - Configuration object.
+ * @param {function=} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {object} callback.instance - The instance.
+ * @param {object} callback.apiResponse - The full API response.
+ */
+ServiceObject.prototype.create = function(options, callback) {
+  var self = this;
+  var args = [this.id];
+
+  if (is.fn(options)) {
+    callback = options;
+  }
+
+  if (is.object(options)) {
+    args.push(options);
+  }
+
+  // Wrap the callback to return *this* instance of the object, not the newly-
+  // created one.
+  function onCreate(err, instance, apiResponse) {
+    if (err) {
+      callback(err, null, apiResponse);
+      return;
+    }
+
+    self.metadata = instance.metadata;
+
+    callback(null, self, apiResponse);
+  }
+
+  args.push(onCreate);
+
+  this.createMethod.apply(null, args);
+};
+
+/**
+ * Delete the object.
+ *
+ * @param {function=} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {object} callback.apiResponse - The full API response.
+ */
+ServiceObject.prototype.delete = function(callback) {
+  var reqOpts = {
+    method: 'DELETE',
+    uri: ''
+  };
+
+  ServiceObject.prototype.request.call(this, reqOpts, function(err, resp) {
```

Comment:
Actually, **get this!** It's because Compute has it's own `request` method; Compute API calls generally receive an Operation response, and we return an Operation object. To simplify all of that, we overwrote `request` instead of just using the default behavior from ServiceObject. Calling the ServiceObject.request method insures we don't get caught up in any custom implementations that were layered on top.